src/combinatorics/simple_graph/degree_sum.lean

```diff
@@ -0,0 +1,244 @@
+@[simps]
+def dart_equiv_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
+{ to_fun := λ d, ⟨d.fst, d.snd, d.is_adj⟩,
+  inv_fun := λ s, ⟨s.fst, s.snd, s.snd.property⟩,
+  left_inv := λ d, by ext; simp,
+  right_inv := λ s, by ext; simp }
+
+instance dart.fintype [fintype V] [decidable_rel G.adj] : fintype G.dart :=
+fintype.of_equiv _ G.dart_equiv_sigma.symm
+
+instance dart.inhabited [inhabited V] [inhabited (G.neighbor_set (default _))] :
+  inhabited G.dart := ⟨G.dart_equiv_sigma.symm ⟨default _, default _⟩⟩
+
+variables {G}
+
+/-- The edge associated to the dart. -/
+def dart.edge (d : G.dart) : sym2 V := ⟦(d.fst, d.snd)⟧
+
+@[simp] lemma dart.edge_mem (d : G.dart) : d.edge ∈ G.edge_set :=
+d.is_adj
+
+/-- Reverses the orientation of a dart. -/
+def dart.rev (d : G.dart) : G.dart :=
+⟨d.snd, d.fst, G.sym d.is_adj⟩
+
+@[simp] lemma dart.rev_edge (d : G.dart) : d.rev.edge = d.edge :=
+sym2.eq_swap
+
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+dart.ext _ _ rfl rfl
+
+@[simp] lemma dart_rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
+dart.rev_rev
+
+lemma dart.rev_ne (d : G.dart) : d.rev ≠ d :=
+begin
+  cases d with f s h,
+  simp only [dart.rev, not_and, ne.def],
+  rintro rfl,
+  exact false.elim (G.loopless _ h),
+end
+
+lemma dart_edge_eq_iff (d₁ d₂ : G.dart) :
+  d₁.edge = d₂.edge ↔ d₁ = d₂ ∨ d₁ = d₂.rev :=
+begin
+  cases d₁ with s₁ t₁ h₁,
+  cases d₂ with s₂ t₂ h₂,
+  simp only [dart.edge, dart.rev_edge, dart.rev],
+  rw sym2.eq_iff,
+end
+
+variables (G)
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_of_neighbor_set (v : V) (w : G.neighbor_set v) : G.dart :=
+⟨v, w, w.property⟩
+
+lemma dart_of_neighbor_set_injective (v : V) : function.injective (G.dart_of_neighbor_set v) :=
+λ e₁ e₂ h, by { injection h with h₁ h₂, exact subtype.ext h₂ }
+
+section degree_sum
+variables [fintype V] [decidable_rel G.adj]
+
+lemma dart_fst_fiber [decidable_eq V] (v : V) :
+  univ.filter (λ d : G.dart, d.fst = v) = univ.image (G.dart_of_neighbor_set v) :=
+begin
+  ext d,
+  simp only [mem_image, true_and, mem_filter, set_coe.exists, mem_univ, exists_prop_of_true],
+  split,
+  { rintro rfl,
+    exact ⟨_, d.is_adj, dart.ext _ _ rfl rfl⟩, },
+  { rintro ⟨e, he, rfl⟩,
+    refl, },
+end
+
+lemma dart_fst_fiber_card_eq_degree [decidable_eq V] (v : V) :
+  (univ.filter (λ d : G.dart, d.fst = v)).card = G.degree v :=
+begin
+  have hh := card_image_of_injective univ (G.dart_of_neighbor_set_injective v),
+  rw [finset.card_univ, card_neighbor_set_eq_degree] at hh,
+  rwa dart_fst_fiber,
+end
+
+lemma dart_card_eq_sum_degrees : fintype.card G.dart = ∑ v, G.degree v :=
+begin
+  haveI h : decidable_eq V := by { classical, apply_instance },
+  simp only [←card_univ, ←dart_fst_fiber_card_eq_degree],
+  exact card_eq_sum_card_fiberwise (by simp),
+end
+
+variables [decidable_eq V]
+
+lemma dart_edge_fiber (d : G.dart) :
```

Comment:
```suggestion
lemma dart.edge_fiber (d : G.dart) :
```
So that `d.edge_fiber` works

Reply:
That's a reasonable suggestion, but I'm torn because the core of the theorem is that `dart_vert_fiber` and `dart_edge_fiber` are meant to identify to fiberings of the `dart` type, and then by summing cardinalities of the fibers you get the degree-sum formula.  I like the parallelism in the names.

It actually should be that `dart_edge_fiber` takes an edge, but because it's a little awkward building the two-element `finset` of darts that map to one, I made it take a dart instead.  Maybe I should look into whether it's so awkward after all.